### PR TITLE
feat(aws-logs-destinations): add Firehose Log Subscription Destination

### PR DESCRIPTION
**Pull request by yerzhan7**
**Add FIrehose CW Logs Subscription Destination** for delivering logs from CW directly into Firehose (in batches, gziped).

Currently, this package contains only Kinesis Logs Subscription Destination, but it's possible to send directly to Firehose without Kinesis stream. See https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html#FirehoseExample

Implementation is based on these official instructions.

~~Also, Firehose does not have high level CDK constructs, so I used native Cfn level stuff.~~

Tested by deploying test stack and sending logs to the log stream. Validated that logs are delivered to S3 via Firehose.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by BenChaimberg** [comment]
Thank you very much for your contribution (especially the documentation)! We are currently working on the design for the Firehose L2 (see #10810) which should be used here. To avoid rewriting the API in the future, I am going to mark this as blocked for now. Please reach out here if you have questions!

File: packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts:9]: Please add some steps for verifying the resources were wired up correctly (see: [CONTRIBUTING#integration-tests](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#integration-tests)). Something like:
- `aws firehose put-record ...`
- `aws get-log-record ...`

File: packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts:101]: Please change this to 1 so the build will fail until this has been fixed.

**Review by BenChaimberg — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts]: Any tips on what might be the issue here? I have no idea...

File: packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts:9]: Done, thanks

File: packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts]: Done

File: packages/@aws-cdk/aws-logs-destinations/README.md
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/README.md:14]: Will wait for L2

File: packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts]: Fixed, thanks

File: packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts]: This was copied from the official docs, but I tested this without it and it looks like that it's not required. 

Removed in R2

**Comment by BenChaimberg** [comment]
In the future, please add more commits onto your working branch instead of amending the old commit and force-pushing. This makes it easier to see what you've changed :)

**Comment by yerzhan7** [comment]
> In the future, please add more commits onto your working branch instead of amending the old commit and force-pushing. This makes it easier to see what you've changed :)

Will do, thanks. I presumed github can do diffs between revisions.

So, do I still need to squash commits after approval?

File: packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/test/kinesisfirehose.test.ts]: But in my first revision it was not working when I used scope of `Stack`.

My initial intention was to have a single role for each Firehose destination. (i.e. 2 log groups will re-use the same role). What scope is the correct one to use for this role?

**Comment by BenChaimberg** [comment]
One would hope but doesn't seem to be the case...or I just haven't found the right buttons to click. No need to squash or manually merge – our workflows will take care of that.

**Comment by yerzhan7** [comment]
@BenChaimberg any ETA for Firehose L2? I saw https://github.com/aws/aws-cdk/pull/14860 got closed.

Is it possible to merge this PR without Firehose L2?

Thanks

**Review by BenChaimberg — DISMISSED** [review]
we're almost there! #15544 is nearing completion and we can restart this work when it's complete

requesting changes just to get this off my queue

**Comment by BenChaimberg** [comment]
Hello! #15544 has been merged and the DeliveryStream L2 is ready and raring to go! Please go take a look – it should have all we need to complete this PR.

**Comment by yerzhan7** [comment]
Awesome! Thanks, will post an update in a few days

File: packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts:35]: CloudWatch Logs currently has a bug where destination ARNs that contain the string "destination" are rejected from creating a subscription. Workaround is to change the stack name below to remove the "destination" – maybe replace with "dests"

Internal reference: P48761216

**Review by BenChaimberg — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts:24]: With your suggested change we would create 2 identical IAM roles that will fail my unit test in https://github.com/aws/aws-cdk/pull/14918/files#diff-4dab4ab34393d725389edf0c2c55b924e6e9c569d86f78b21ad5d36b8f29b076R104

Do we really want to create 2 identical roles in that unit test case?

File: packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts
Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/test/integ.kinesisfirehose-to-s3.ts:35]: What a weird bug! Thanks for digging into this.

File: packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts:23]: I think we do. There should be a logical mapping between the actor and the role, which would be violated if we create one role per delivery stream (as that is the resource). The test case can be modified to demonstrate that we re-use the role if there are multiple Firehose subscriptions from the same log group

Comment (yerzhan7) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts:23]: > The test case can be modified to demonstrate that we re-use the role if there are multiple Firehose subscriptions from the same log group

This would still create 2 separate roles because `scope` would be different for 2 separate subscriptions even if they are applied to the same log group.

So, it looks like `scope.node.tryFindChild(roleId) as iam.IRole ??` is a redundant check as that condition will never be true.

Anyway, I will just remove this unit test case for now as it's impossible to test that condition.

Comment (BenChaimberg) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts]: Ah because the destination is bound under SubscriptionFilter, not LogGroup. Okay, no problem with having multiple roles then

Comment (BenChaimberg) [reply packages/@aws-cdk/aws-logs-destinations/lib/kinesisfirehose.ts]: To be clear, yes we should remove the `tryFindChild` and always create a new role

**Comment by yerzhan7** [comment]
@BenChaimberg , could you please help me with CodeBuild failure above? Thanks

```
aws-cdk-lib: [96mlib/aws-logs-destinations/lib/kinesisfirehose.ts[0m:[93m11[0m:[93m59[0m - [91merror[0m[90m TS2694: [0mNamespace '"/codebuild/output/src850447146/src/github.com/aws/aws-cdk/packages/aws-cdk-lib/lib/aws-kinesisfirehose/index"' has no exported member 'IDeliveryStream'.
aws-cdk-lib: [7m11[0m     constructor(private readonly deliveryStream: firehose.IDeliveryStream) {
```

**Comment by BenChaimberg** [comment]
aws-cdk-lib is the v2 library which removes experimental modules (except for the L1s), meaning stable modules that are included cannot currently depend on experimental modules. The team is currently working on this issue so we may have a fix in the pipeline, but this PR will be blocked until then.

**Comment by nikovirtala** [comment]
@BenChaimberg, is there anything that one could do to get this merged? – The issue with experimental modules with v2 is, in my understanding, solved.

**Comment by yerzhan7** [comment]
> The issue with experimental modules with v2 is, in my understanding, solved.

@nikovirtala I've just merged latest stuff from master and, as you see above, the build still fails with the same error message.

How do you know that the issue solved? Maybe I am missing some config

**Comment by BenChaimberg** [comment]
@kaizen3031593 can you please respond to the query above -- I have not been following the developments on experimental modules.

**Comment by kaizen3031593** [comment]
@nikovirtala my quick understanding of the issue is that stable modules cannot depend on experimental modules; this is by design. In that sense, there is nothing to resolve. Instead, we are waiting for `kinesis-firehose` to be stabilized, in which case we can take another look at this PR. I will double-check for you, but if you don't hear otherwise that is the signal we are looking for.

**Comment by nikovirtala** [comment]
> @nikovirtala my quick understanding of the issue is that stable modules cannot depend on experimental modules; this is by design. In that sense, there is nothing to resolve. Instead, we are waiting for `kinesis-firehose` to be stabilized, in which case we can take another look at this PR. I will double-check for you, but if you don't hear otherwise that is the signal we are looking for.

It is not an option to release these changes as experimental too?

**Comment by kaizen3031593** [comment]
@nikovirtala we can release changes as experimental in a stable module by using the `beta` prefix (i.e. `myExperimentalConstruct-Beta`). However what we are talking about is having a stable module _depend_ on an experimental one, which is a different ballgame (and not possible). It is possible to release this as an entirely new experimental construct that can then depend on both experimental and stable modules, but that is overkill here. I think it makes more sense to wait for `kinesis-firehose` to stabilize.

I am going to remove my assignment from this for now but feel free to ping me if you want my attention. When `kinesis-firehose` is stable I will revisit this PR.

**Pull request closed without merging**
